### DISCUSSION TITLE
Edit real-time component without form submit and using 2 related models and validation

### DISCUSSION THREAD

**Discussion by antmer** [livewire/livewire#3580] [1. Help] [answered]
Hi everyone,
I'm new with Livewire

I explain my problem:

I have 2 models : 
- Product (id, label, total_material_cost, reduction_percentage, reduction_cost, total_product_cost) 
Product hasMany(Material)
- Material (id, product_id, label, price, multiple1, multiple2, total_price_with_multiple)

I would like to create a page which look like an excel page (every time I change a number value, the rest of the page is calculated). For that reason I don't want a submit button. But I need validation.
When I change an input with wire:model, I have to execute the calculateProduct() function to actualize all Product properties before render.

I don't understand how to do that with Livewire...
I first try with "wire:change" for each input, but it doesn't allow me to use validation rules
```blade
<input type="text" wire:change="saveMaterial($event.target.value, 'price', {{ $material->id }})" value="{{ $material->price }}" />
```

Here is an example for my page :

```blade
<table>
    <thead>
        <tr>
            <th></th>
            <th>Label</th>
            <th>Price</th>
            <th>Multiple 1</th>
            <th>Multiple 2</th>
            <th>Total</th>
        </tr>
    </thead>
    <tbody>
        @foreach($product->materials as $index => $material)
        <tr>
            <td>
                <x-confirm-delete-form />
            </td>
            <td>
                <input type="text" wire:model="label"/>
            </td>
            <td>
                <input type="text" wire.model="price" />
            </td>
            <td>
                <input type="text" wire:model="multiple1" />
            </td>
            <td>
                <input type="text" wire:model="multiple2" />
            </td>
            <td class="total">
                {{ $total_price_with_multiple }}
            </td>
        </tr>
        @endforeach
        <tr class="ensemble">
            <th colspan="4"></th>
            <th>ENS</th>
            <th>{{ $product->total_material_cost }}</th>
        </tr>
    </tbody>
</table>

<!--I can add as many Material as I need for my Product-->
<button>
    Add one material
</button>

<!--REDUCTION & TOTAL_PRODUCT_COST-->
<table>
    <tbody>
        <tr>
            <td>Reduction %</td>
            <td><input type="text"  wire:model="reduction_percentage" placeholder="Decimal, e.g. 0.05" /></td>
            <td></td>
            <td>{{ $product->reduction_cost }}</td>
        </tr>
        <tr>
            <td colspan="2"></td>
            <th>Total Cost:</th>
            <th>{{ $product->total_product_cost }}</th>
        </tr>
    </tbody>
</table>
```

I hope it is clear
Thank you a lot for your help.

**Comment by joshhanley** [comment, marked as answer]
@antmer what you are after is called the "updated" hook. You can see it's use here, but instead of using `$this->validateOnly` like in the example, instead you will use `$this->validate()` https://laravel-livewire.com/docs/2.x/input-validation#real-time-validation

Then after your validation is run, then you can call your calculation function, whatever that may be 🙂

This will require you to change all of your inputs back to `wire:model` though.

For more details on hooks, check out https://laravel-livewire.com/docs/2.x/lifecycle-hooks

Hope this helps!

**Reply by antmer** [reply]
Hi Josh,

Thank you it helps me! 😊

But I have another problem...
I can't bind the $product->materials in my component.
I have this :
```php
class ProductEdit extends Component
{
    public $product;
    public $materials;
    
    protected $rules=[
        'materials.*.label' => 'required',
        'materials.*.price' => 'required',
        'materials.*.multiple1' => 'required',
        'materials.*.multiple2' => 'required',
    ];


    public function render()
    {
        $this->product = Product::with('materials')->find($this->product->id); //with() is for eager-loading
        $this->materials = $this->product->materials ;

        return view('livewire.work-edit');
    }
}

```
And the view :
```blade
...
        @foreach($product->materials as $index => $material)
        <tr>
            ...
            <td>
                <input type="text" wire:model.lazy="materials.{{ $index }}.price"/>
            </td>
            ...
        </tr>
        @endforeach
...
```
All of the material lines are displayed in the foreach but I have no value in the input fields.
Do you have any idea?

**Reply by joshhanley** [reply]
@antmer hmm not sure exactly why that wouldn't be working. But in your foreach, as you already have a materials public property, change it to use that directly

```blade
@foreach($materials as $index => $material)
```

If you have Livewire devtools in chrome, it can help see what data is being sent to the front end, which will then populate the fields. Maybe check to see if the `$materials` property has the correct values being sent to the front end.

Hope this helps!

**Reply by antmer** [reply]
Just a careless mistake... I forgot the colon on "wire:model" (not on the example above but in my code)
And I have downloaded the extension. Useful!

Thank you for your help! 😃

**Reply by joshhanley** [reply]
@antmer ah yep it happens! No worries 🙂
